### PR TITLE
Fix build for non-GOPATH users

### LINKED ISSUE #28890: Unable to build kubernetes running 'make'
After merging https://github.com/kubernetes/kubernetes/pull/25978, I am unable to build kubernetes:

```sh
./hack/build-go.sh
NOTE: ./hack/build-go.sh has been replaced by 'make' or 'make all'

The equivalent of this invocation is: 
    make 


make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.
```

Before ``eecbfb1a2808b233a4c7032ce26e260276bffde5``:
```sh
./hack/build-go.sh
Go version: go version go1.6.2 linux/amd64
+++ [0713 07:36:27] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0713 07:36:27] Building go targets for linux/amd64:
    cmd/kube-dns
...
```

Am I something missing?

### LINKED ISSUE #28987: make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.
No problem until yesterday, now I tried again many times, including clone again, but problem exists:

root@ubuntu:/home/zte/share/k8s-clone/kubernetes# git commit
Checking that it builds... 
make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.

Checking for docs that need updating... make: *** No rule to make target `/*.go', needed by `_output/bin/deepcopy-gen'.  Stop.
!!! Error in hack/verify-munge-docs.sh:29
  'make -C "${KUBE_ROOT}/" WHAT=cmd/mungedocs' exited with status 2
Call stack:
  1: hack/verify-munge-docs.sh:29 main(...)
Exiting with status 1
ERROR!
Some docs are out of sync between CLI and markdown.
To regenerate docs, run:
  hack/update-munge-docs.sh

Checking for swagger type documentation that need updating... ERROR!
Swagger type documentation needs to be updated.
To regenerate the spec, run:
  hack/update-generated-swagger-docs.sh

  1: /home/zte/share/k8s-clone/kubernetes/hack/lib/golang.sh:587 kube::golang::build_binaries(...)
  2: hack/make-rules/build.sh:26 main(...)
Exiting with status 1
make: *** [all] Error 1
!!! Error in hack/verify-swagger-spec.sh:26
  'make -C "${KUBE_ROOT}" WHAT=cmd/kube-apiserver' exited with status 2
Call stack:
  1: hack/verify-swagger-spec.sh:26 main(...)
Exiting with status 1
ERROR!
Swagger spec needs to be updated.
To regenerate the spec, run:
  hack/update-swagger-spec.sh

Aborting commit

### PR DESCRIPTION
**Pull request by thockin**
I want feedback on this.  There are 2 commits here, we should probably only do one of them.  The release note will depend on which one we choose.  Also docs need update.

Fixes #28890
Fixes #28987

### CODE REVIEW

**Comment by lavalamp** [comment]
I am a big fan of the second approach (demand user set up a GOPATH). It will cause a small amount of pain for some people, but in the long run, your life is 100x easier if you make a separate GOPATH for each go project you want to work with, to keep the dependencies nicely separated.

**Comment by thockin** [comment]
NB: Demanding that they be in a GOPATH does not imply demanding a
one-element GOPATH.  We could do that, too, but that was not what was
discussed.  @lavalamp this touches on your own GOPATH mess - the first
commit here gives us a chance to reliably manipulate GOPATH.  The second
does not.

**Comment by aledbf** [comment]
@thockin since #25978 I only get errors running `make`.
(I used the second commit)
```
old-mbp:kubernetes aledbf$ pwd
/Users/aledbf/go/src/k8s.io/kubernetes
old-mbp:kubernetes aledbf$ go version
go version go1.6.2 darwin/amd64
old-mbp:kubernetes aledbf$ go env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/aledbf/go"
GORACE=""
GOROOT="/usr/local/Cellar/go/1.6.2/libexec"
GOTOOLDIR="/usr/local/Cellar/go/1.6.2/libexec/pkg/tool/darwin_amd64"
GO15VENDOREXPERIMENT="1"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common"
CXX="clang++"
CGO_ENABLED="1"
```
make
```
old-mbp:kubernetes aledbf$ make
make: *** No rule to make target `\nt/Users/aledbf/go/src/k8s.io/kubernetes/cmd/libs/go2idl/deepcopy-gen/main.go', needed by `_output/bin/deepcopy-gen'.  Stop.
```

with debug
````
old-mbp:kubernetes aledbf$ DBG_MAKEFILE=1 make
Makefile:17: ***** starting makefile for goal(s) ""
Makefile:18: ***** miércoles, 13 de julio de 2016, 13:50:32 CLT
Makefile.generated_files:93: ***** finding all *.go dirs
Makefile.generated_files:160: ***** finding all +k8s: tags
Makefile.generated_files:188: ***** finding all +k8s:deepcopy-gen tags
Makefile.generated_files:296: ***** finding all +k8s:conversion-gen tags
mkdir -p .make/_output/bin;                                        \
	(echo -n "_output/bin/conversion-gen: ";                        \
	 DIRECT=$(go list -e -f '{{.Dir}} {{.Dir}}/*.go' ./cmd/libs/go2idl/conversion-gen);                \
	 INDIRECT=$(go list -e   -f '{{range .Deps}}{{.}}{{"\n"}}{{end}}' ./cmd/libs/go2idl/conversion-gen | grep "^k8s.io/kubernetes" | xargs go list -e -f '{{.Dir}} {{.Dir}}/*.go');  \
	 echo $DIRECT $INDIRECT | sed 's/ / \\\n\t/g';       \
	) | sed "s|$(pwd -P)/||" > .make/_output/bin/conversion-gen.mk.tmp;                    \
	cmp -s .make/_output/bin/conversion-gen.mk.tmp .make/_output/bin/conversion-gen.mk || cat .make/_output/bin/conversion-gen.mk.tmp > .make/_output/bin/conversion-gen.mk && rm -f .make/_output/bin/conversion-gen.mk.tmp
mkdir -p .make/_output/bin;                                        \
	(echo -n "_output/bin/deepcopy-gen: ";                          \
	 DIRECT=$(go list -e -f '{{.Dir}} {{.Dir}}/*.go' ./cmd/libs/go2idl/deepcopy-gen);                  \
	 INDIRECT=$(go list -e   -f '{{range .Deps}}{{.}}{{"\n"}}{{end}}' ./cmd/libs/go2idl/deepcopy-gen | grep "^k8s.io/kubernetes" | xargs go list -e -f '{{.Dir}} {{.Dir}}/*.go');  \
	 echo $DIRECT $INDIRECT | sed 's/ / \\\n\t/g';       \
	) | sed "s|$(pwd -P)/||" > .make/_output/bin/deepcopy-gen.mk.tmp;                    \
	cmp -s .make/_output/bin/deepcopy-gen.mk.tmp .make/_output/bin/deepcopy-gen.mk || cat .make/_output/bin/deepcopy-gen.mk.tmp > .make/_output/bin/deepcopy-gen.mk && rm -f .make/_output/bin/deepcopy-gen.mk.tmp
hack/make-rules/precheck-gopath.sh
make: *** No rule to make target `\nt/Users/aledbf/go/src/k8s.io/kubernetes/cmd/libs/go2idl/deepcopy-gen/main.go', needed by `_output/bin/deepcopy-gen'.  Stop.
```

**Comment by lavalamp** [comment]
I did not mean to insist on a one-element GOPATH (indeed my scheming about methodology for splitting components into separate repos is going to involve multi-element GOPATHs). Rather, insist that it's in GOPATH, and recommend in the failure message that the developer maintain separate GOPATHs for each go project, to avoid dependency cross contamination.

**Comment by eparis** [comment]
I don't love `Demand that builds run in a GOPATH` because it seems to prevent the typical
- download tar.gz
- un-tar-gz
- cd kubernetes
- make
Which I'd love to see keep working...

**Comment by mml** [comment]
I also don't like that I'll have to correct my environment depending on my working directory.  I can do this with zsh tricks like using `chpwd()` and/or `preexec()`, but that is a big burden on more casual contributors.

**Comment by mml** [comment]
Another point: if we rely on the user-set GOPATH, the user could very easily be compiling the wrong code without realizing it.  I think the normal expectation is that barring some explicit request to do otherwise, running 'make' in a directory makes *that stuff*.

http://stackoverflow.com/questions/18136918/how-to-get-current-relative-directory-of-your-makefile is a (GNU) makefile recipe that would allow us to construct a proper GOPATH in the makefile if not otherwise given one via some other explicit override (e.g. `make KUBEGOPATH=blah`)

**Comment by lavalamp** [comment]
I guess the question is, are we going to satisfy the expectations of the typical 'make' user or of the typical golang user?

**Comment by eparis** [comment]
@lavalamp I'm not sure how the first patch breaks the typical 'go' user, but I haven't read the whole issue either, it have tons of cases. Can I get the ELI5 version?

**Comment by lavalamp** [comment]
The bad part of the first patch is that I, and my tools, cannot predict what code a given import path is going to import, because I don't know what sort of magic GOPATH the makefile is going to assemble for me.

**Comment by spxtr** [comment]
I agree with requiring `GOPATH`. We're a go project, and in order to make all the go tools work we need to obey [How to Write Go Code](https://golang.org/doc/code.html). Go developers should be familiar with this already, and new developers should familiarize themselves with the go tools.

In the path we've mucked with fake `GOPATH`s, and this lead to strange bugs and developer confusion.

**Comment by mml** [comment]
> The bad part of the first patch is that I, and my tools, cannot predict what code a given import path is going to import, because I don't know what sort of magic GOPATH the makefile is going to assemble for me.

OK, agreed.  I have experienced this pain several times.  For example trying to run my own syntax checker or other tool from inside vim.

**Comment by thockin** [comment]
@aledbf that looks like a mac-vs-PC problem.  Is it actually mis-processing end-of-lines?

**Comment by thockin** [comment]
I am sympathetic to both approaches (hence I implemented both).

I agree that mucking with GOPATH is annoying, and that has always been my big gripe with Go.  But it is how Go works.

On the other hand, setting up our own private gopath automatically is clean and guarantees that no external deps leak in.  But it is ugly and hard to debug and has little turdlets all over the codebase...

**Comment by soltysh** [comment]
I've checked the `Demand that builds run in a GOPATH` with `make clean all` and it not working as previously :(
```
$ make clean all
+++ [0714 16:37:18] Verifying Prerequisites....
!!! [0714 16:37:19] Build image not built.  Cannot clean via docker build image.
+++ [0714 16:37:19] Removing data container kube-build-data-0bc7d68a56
+++ [0714 16:37:19] Removing _output directory
+++ [0714 16:37:19] Deleting docker image kube-build:build-0bc7d68a56
+++ [0714 16:37:19] Cleaning all other untagged docker images
Go version: go version go1.7rc1 linux/amd64
+++ [0714 16:37:19] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0714 16:37:19] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [0714 16:37:20] Placing binaries
make: *** No rule to make target '.make/federation/apis/federation/gofiles.mk.stamp', needed by 'federation/apis/federation/zz_generated.deepcopy.go'.  Stop.
```
But I still vote for that demand.

**Comment by ncdc** [comment]
I've tested the 2nd commit in f24. I'm able to do something like `make WHAT=...` just fine the first time. But if I then try to run `sudo hack/local-up-cluster.sh`, I get stopped by the gopath precheck. It appears that something is clearing my session's `GOPATH` for some reason.

**Comment by ncdc** [comment]
Nevermind, sudo fun :)

File: hack/make-rules/precheck-gopath.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# Make sure the current directory is inside a GOPATH element.
+parts="$(echo "${GOPATH:-}" | sed 's/:/ /g')"
```
Comment (ncdc) [inline hack/make-rules/precheck-gopath.sh]: I'd recommend exiting 1 if GOPATH is unbound. Otherwise the error message is a bit unclear (you have to build from a valid GOPATH and the current dir is in what you think you have for GOPATH, but it's empty (sudo invocation))

**Comment by thockin** [comment]
Based on some conversations about how to break up the repo over the next few months, it's pretty clear that making a private GOPATH is a necessary evil.  So I think the first commit here has to stand.

The question then is whether we still need to demand that CWD is under GOPATH, when in fact we don't require that.  I'm going to split this PR into 2 parts to make progress.

**Comment by thockin** [comment]
@lavalamp, I think we should proceed with this PR.  Disagree?

**Comment by dims** [comment]
@thockin Looks good to me 👍

**Comment by spxtr** [comment]
> Based on some conversations about how to break up the repo over the next few months, it's pretty clear that making a private GOPATH is a necessary evil. So I think the first commit here has to stand.

Why is it a necessary evil? I would much rather give users two options:

1. Use their own GOPATH. This means they might have a polluted build but that might be what they want anyway, such as if they are adding a new dependency. This gives fast, iterative builds and allows use of the go tools with no shim shell scripts.
1. Do the entire build in a fresh container with the source mounted at `/go/k8s.io/kubernetes`.

This way there is no need to much with setting up a private GOPATH on the user's machine. Just mounting the volume to the container at the right location is enough.

**Comment by thockin** [comment]
@spxtr to bridge with the evolution of a fully-extracted and `go get` friendly client, the plan involves abusing GOPATH.  This means we'll be back to relying on our setup anyway.

Once we have a full dockerized build, we can encapsulate this inside the docker run, but for now I think it gives the most complete solution.

**Comment by thockin** [comment]
@dims said LGTM

**Comment by lavalamp** [comment]
Yeah, I agree. We need a facility for setting GOPATH so tools will work.
maybe `source $(make gopath)` or some such? You're the make expert. :)
